examples/http-server/main/main.c

```diff
@@ -0,0 +1,147 @@
+#include "server.h"
+#include "esp_http_server.h"
+#include "epd.h"
+#include "esp_heap_caps.h"
+
+#define WRITE_HEADER(req, buffer, name, format, src) \
+    sprintf(buffer, format, src);                    \
+    ESP_ERROR_CHECK(httpd_resp_set_hdr(req, name, buffer));
+static esp_err_t http_index(httpd_req_t* req) {
+    EpdData data = n_epd_data();
+    char width[20], height[20], temperature[20];
+    WRITE_HEADER(req, width, "width", "%d", data.width);
+    WRITE_HEADER(req, height, "height", "%d", data.height);
+    WRITE_HEADER(req, temperature, "temperature", "%d", data.temperature);
+    const char* response = "Hello! Check headers\n";
+    httpd_resp_set_type(req, "text/plain");
+    httpd_resp_set_status(req, "200");
+    httpd_resp_send(req, response, HTTPD_RESP_USE_STRLEN);
+    return ESP_OK;
+}
+
+esp_err_t http_clear(httpd_req_t* req) {
+    ESP_LOGI(__FUNCTION__, "Clear\n");
+    n_epd_clear();
+    const char* response = "Cleared\n";
+    httpd_resp_set_type(req, "text/plain");
+    httpd_resp_set_status(req, "200");
+    httpd_resp_send(req, response, HTTPD_RESP_USE_STRLEN);
+    return ESP_OK;
+}
+
+esp_err_t http_draw(httpd_req_t* req) {
+    // optional headers: x,y. Default to 0
+    // required headers: height, width
+    // Content should be a stream of special bytes - we're reading 4 bits at a time.
+    int x, y, width, height, clear;
+    char header[20];
+    memset(header, 0, 20);
+    if (httpd_req_get_hdr_value_str(req, "clear", header, 20) == ESP_OK) {
+        sscanf(header, "%d", &clear);
+    } else {
+        clear = 0;
+    }
+    if (httpd_req_get_hdr_value_str(req, "x", header, 20) == ESP_OK) {
+        sscanf(header, "%d", &x);
+    } else {
+        x = 0;
+    }
+    if (httpd_req_get_hdr_value_str(req, "y", header, 20) == ESP_OK) {
+        sscanf(header, "%d", &y);
+    } else {
+        y = 0;
+    }
+    if (httpd_req_get_hdr_value_str(req, "width", header, 20) == ESP_OK) {
+        sscanf(header, "%d", &width);
+    } else {
+        char response[60];
+        sprintf(response, "Missing header width");
+        httpd_resp_set_type(req, "text/plain");
+        httpd_resp_set_status(req, "400");
+        httpd_resp_send(req, response, HTTPD_RESP_USE_STRLEN);
+        return ESP_OK;
+    }
+    if (httpd_req_get_hdr_value_str(req, "height", header, 20) == ESP_OK) {
+        sscanf(header, "%d", &height);
+    } else {
+        char response[60];
+        sprintf(response, "Missing header height");
+        httpd_resp_set_type(req, "text/plain");
+        httpd_resp_set_status(req, "400");
+        httpd_resp_send(req, response, HTTPD_RESP_USE_STRLEN);
+        return ESP_OK;
+    }
+
+    // READING STREAM
+    int req_size = req->content_len;
+    char* content = (char*)heap_caps_malloc(req_size, MALLOC_CAP_SPIRAM);
```

Comment:
This memory buffer is not deallocated after receiving data, which could cause a memory leak.